### PR TITLE
Add new package: parabricks/GenomeWorks

### PR DESCRIPTION
**Pull request by dunatotatos**
The dependencies match the requirements on the GitHub repo: https://github.com/clara-parabricks/GenomeWorks
Except CUDA which is bumped to @11, for 2 reasons:
- Cuda <11 can only be compiled with gcc@4.8.5 on RHEL 7, but GenomeWorks requires gcc@5.4.0:
- the provided patch to ignore CUB only works with Cuda@11

### CODE REVIEW

**Comment by dunatotatos** [comment]
Not sure why linux tests / unittests (3.7) failed. If someone can help me understanding, I'm happy to fix the PR.

**Comment by alalazo** [comment]
@dunatotatos I restarted the tests. We have a unit test that checks a few aspects of process spawning that is "non-deterministic" and so sometimes gives false negatives.

File: var/spack/repos/builtin/packages/genomeworks/package.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    depends_on('cmake@3.10.2:', type=('build'))
+    depends_on('cuda@11:', type=('build', 'run'))
+    depends_on('python@3.6.7:', type=('build', 'run'))
+
+    phases = ['cmake', 'install']
+    patch('3rdparty.patch')
+
+    def cmake_args(self):
+        args = ["-DCMAKE_BUILD_TYPE=Release"]
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/genomeworks/package.py]: Spack already adds this flag as part of the base package. Does the default work? If not, you can override the variant like other packages do.

File: var/spack/repos/builtin/packages/genomeworks/package.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    version('0.4.3', tag='v0.4.3', submodules=True)
+    version('0.4.0', tag='v0.4.0', submodules=True)
+    version('0.3.0', tag='v0.3.0', submodules=True)
+    version('0.2.0', tag='v0.2.0', submodules=True)
+
+    depends_on('cmake@3.10.2:', type=('build'))
+    depends_on('cuda@11:', type=('build', 'run'))
+    depends_on('python@3.6.7:', type=('build', 'run'))
+
+    phases = ['cmake', 'install']
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/genomeworks/package.py]: Do the default phases (cmake/build/install) not work?

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 1 above)
Comment (dunatotatos) [reply var/spack/repos/builtin/packages/genomeworks/package.py]: The default does not work, as I need to add flags for CUDA (line 37 and below). I therefore checked the first 20 packages in var/spack/repos/builtin/. Either the cmake args are not changed, or the whole cmake_args function is re-defined. I also read this way of working in the documentation [here](https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake). Is that correct?

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 2 above)
Comment (dunatotatos) [reply var/spack/repos/builtin/packages/genomeworks/package.py]: I don't know if the default works, but I'm following the instructions from https://github.com/clara-parabricks/GenomeWorks `make` is not mentioned, and I don't want this package to break because I assumed that `make` should be part of the installation process. I prefer to implement the exact documented steps. Is that correct?

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/genomeworks/package.py]: @dunatotatos This is not what @adamjstewart meant, I guess. The `cmake_args` method is used here in Spack:

https://github.com/spack/spack/blob/cfd69da8391dbe6acc38ae29274ba3b48bb81ba3/lib/spack/spack/build_systems/cmake.py#L333-L339

The most commonly used options (i.e. which generator to use and the build type) are gathered in the `std_cmake_args` property, therefore there's no need to set `-DCMAKE_BUILD_TYPE=Release`.

**TL;DR** I think Adam meant to just delete line 35, not to use the default `cmake_args` and delete the method completely.

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 2 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/genomeworks/package.py]: `phases` are not meant to be overridden in the majority of cases, since they are specific to each build system. Usually in CMake `make install` in the generated Makefiles implies a `make`, therefore the default will do. Can you delete this line too?

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 1 above)
Comment (dunatotatos) [reply var/spack/repos/builtin/packages/genomeworks/package.py]: Ah, understood! I'll update the line to `args = []` and test with that.

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 2 above)
Comment (dunatotatos) [reply var/spack/repos/builtin/packages/genomeworks/package.py]: Testing without. If it works, I'll update the PR.

File: var/spack/repos/builtin/packages/genomeworks/package.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    version('0.4.3', tag='v0.4.3', submodules=True)
+    version('0.4.0', tag='v0.4.0', submodules=True)
+    version('0.3.0', tag='v0.3.0', submodules=True)
+    version('0.2.0', tag='v0.2.0', submodules=True)
+
+    depends_on('cmake@3.10.2:', type=('build'))
+    depends_on('cuda@11:', type=('build', 'run'))
+    depends_on('python@3.6.7:', type=('build', 'run'))
+
+    patch('3rdparty.patch')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/genomeworks/package.py:34]: Can you comment where the patch is coming from and which issue it solves?

**Review by alalazo — CHANGES_REQUESTED** [review]
Only one minor request, other than that LGTM.

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 3 above)
Comment (dunatotatos) [reply var/spack/repos/builtin/packages/genomeworks/package.py:34]: Patch was written by myself, to solve an incompatibility with CUDA 11. The issue is mentioned here: https://github.com/clara-parabricks/GenomeWorks/issues/570 CUDA 11 already provides CUB, which therefore should not be built from GenomeWorks sub-modules.

This patch breaks the installation with CUDA <11, which is anyway forbidden by `depends_on('cuda@11:,`

Comment (alalazo) [reply var/spack/repos/builtin/packages/genomeworks/package.py:34]: Can you add a comment on top of the patch directive mentioning https://github.com/clara-parabricks/GenomeWorks/issues/570 and describing briefly what you say in the reply above?

File: var/spack/repos/builtin/packages/genomeworks/package.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    """SDK for GPU accelerated genome assembly and analysis."""
+
+    homepage = "https://clara-parabricks.github.io/GenomeWorks/"
+    url      = "https://github.com/clara-parabricks/GenomeWorks/archive/v0.5.3.tar.gz"
+    git      = "https://github.com/clara-parabricks/GenomeWorks.git"
+
+    version('0.5.3', tag='v0.5.3', submodules=True)
+    version('0.5.2', tag='v0.5.2', submodules=True)
+    version('0.5.1', tag='v0.5.1', submodules=True)
+    version('0.5.0-rc1', tag='v0.5.0-rc1', submodules=True)
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/genomeworks/package.py]: We don't usually add release candidates, so I would:
```suggestion
```

**Review by alalazo — APPROVED** [review]
LGTM. Let me know if there's a reason to keep the rc version. Otherwise I'll merge after its removal.

File: var/spack/repos/builtin/packages/genomeworks/package.py (hunk 4 above)
Comment (dunatotatos) [reply var/spack/repos/builtin/packages/genomeworks/package.py]: No reason. Removing.
